pkg/costattribution/manager.go

```diff
@@ -0,0 +1,162 @@
+	logger          log.Logger
+	inactiveTimeout time.Duration
+	limits          *validation.Overrides
+
+	mtx                   sync.RWMutex
+	trackersByUserID      map[string]*Tracker
+	reg                   *prometheus.Registry
+	cleanupInterval       time.Duration
+	metricsExportInterval time.Duration
+}
+
+func NewManager(cleanupInterval, exportInterval, inactiveTimeout time.Duration, logger log.Logger, limits *validation.Overrides, reg *prometheus.Registry) (*Manager, error) {
+	m := &Manager{
+		trackersByUserID:      make(map[string]*Tracker),
+		limits:                limits,
+		mtx:                   sync.RWMutex{},
+		inactiveTimeout:       inactiveTimeout,
+		logger:                logger,
+		reg:                   reg,
+		cleanupInterval:       cleanupInterval,
+		metricsExportInterval: exportInterval,
+	}
+
+	m.Service = services.NewTimerService(cleanupInterval, nil, m.iteration, nil).WithName("cost attribution manager")
+	if err := reg.Register(m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *Manager) iteration(_ context.Context) error {
+	return m.purgeInactiveAttributionsUntil(time.Now().Add(-m.inactiveTimeout).Unix())
+}
+
+func (m *Manager) EnabledForUser(userID string) bool {
+	if m == nil {
+		return false
+	}
+	return len(m.limits.CostAttributionLabels(userID)) > 0
+}
+
+func (m *Manager) Tracker(userID string) *Tracker {
+	if !m.EnabledForUser(userID) {
+		return nil
+	}
+
+	// Check if the tracker already exists, if exists return it. Otherwise lock and create a new tracker.
+	m.mtx.RLock()
+	tracker, exists := m.trackersByUserID[userID]
+	m.mtx.RUnlock()
+	if exists {
+		return tracker
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	if tracker, exists = m.trackersByUserID[userID]; exists {
+		return tracker
+	}
+	tracker = newTracker(userID, m.limits.CostAttributionLabels(userID), m.limits.MaxCostAttributionCardinalityPerUser(userID), m.limits.CostAttributionCooldown(userID), m.logger)
+	m.trackersByUserID[userID] = tracker
+	return tracker
+}
+
+func (m *Manager) Collect(out chan<- prometheus.Metric) {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+	for _, tracker := range m.trackersByUserID {
+		tracker.Collect(out)
+	}
+}
+
+func (m *Manager) Describe(chan<- *prometheus.Desc) {
+	// Describe is not implemented because the metrics include dynamic labels. The Manager functions as an unchecked exporter.
+	// For more details, refer to the documentation: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+}
+
+func (m *Manager) deleteTracker(userID string) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	delete(m.trackersByUserID, userID)
+}
+
+func (m *Manager) updateTracker(userID string) *Tracker {
+	t := m.Tracker(userID)
+
+	if t == nil {
+		m.deleteTracker(userID)
+		return nil
+	}
+
+	newTrackedLabels := m.limits.CostAttributionLabels(userID)
+
+	// sort the labels to ensure the order is consistent
+	sort.Slice(newTrackedLabels, func(i, j int) bool {
```

Comment:
Is it safe to modify the slice returned by `m.limits.CostAttributionLabels(userID)`?

Reply:
modify the copy instead, addressed in 888d8b06231e3e67d1cb7ccf38bac0b4275a95c9